File: 129-求根到叶子节点数字之和.py
class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        result = 0
        mylist = []
        mylist.append((root, 0))

        while mylist:
            node, value = mylist.pop()
            if not node:
                continue
            elif not node.left and not node.right:
                result += value * 10 + node.val
            else:
                mylist.append((node.left, value * 10 + node.val))
                mylist.append((node.right, value * 10 + node.val))
        
        return result

File: 129-求根到叶子节点数字之和.cpp
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int result = 0;
        stack<pair<TreeNode*, int>> mystack;
        mystack.push(make_pair(root, 0));
        TreeNode* node;
        int value;

        while(!mystack.empty()){
            node = mystack.top().first;
            value = mystack.top().second;
            mystack.pop();
            if(!node)                      // 只有到叶子节点才处理
                continue;
            else if(!node->left && !node->right)
                result += value * 10 + node->val;
            else{
                mystack.push(make_pair(node->left, value * 10 + node->val));
                mystack.push(make_pair(node->right, value * 10 + node->val));
            }
        }

        return result;
    }
};
